pkg/gcv/cf/audit.go

```diff
@@ -27,6 +25,10 @@ audit[result] {
 
 	asset := inventory[_]
 	constraint := constraints[_]
+	
+	re_match(constraint.spec.match.gcp.target[_], asset.ancestry_path)
+	exclusion_match := {asset.ancestry_path | re_match(constraint.spec.match.gcp.exclude[_], asset.ancestry_path)}
```

Comment:
Is there a "gcp" fragment in there? I don't quite remember.
According to https://github.com/forseti-security/policy-library/blob/master/docs/user_guide.md#instantiate-constraints, there is no "gcp".
But it exists in https://github.com/forseti-security/policy-library/blob/master/samples/always_violates.yaml

Reply:
I believe we went back and forth on this. Since the [Gatekeeper](https://github.com/open-policy-agent/gatekeeper#constraints) format seems to have adopted nesting within the match, I think we could keep `gcp` in there as the wrapper for our logic.

We will need to do some work to converge the documentation to include this. /cc @maxsmythe

Reply:
Hmm.. It looks like there are 3 matchers in Gatekeepr. The `kinds` matcher has nesting, but the `namespaces` and `labelSelector` matchers don't.

With the `namespaces` matcher, it can look like the following:
```
apiVersion: constraints.gatekeeper.sh/v1alpha1
kind: K8sRequiredLabels
metadata:
  name: ns-must-have-gk
spec:
  match:
    namespaces: ["ns1", "ns2", "ns3"]
  parameters:
    labels: ["gatekeeper"]
```

Do we expect to have other matchers in config validator? For example, the current "gcp" matcher looks at resource hierarchy, and we can have the equivalent of `labelSelector` matcher in the future to do label-based targeting.

Reply:
Yeah, I think it's plausible we might implement additional matchers in the future.

Reply:
For GK, `kinds` is a list of JSON objects and labelSelector is a single object that does allow some sub fields that are also lists.

Their should ideally be no "gcp" subfield, as if we ever allow multi-target constraints this would cause repetition (spec.matches.gcp.gcp.* vs spec.matches.gcp).

The nesting within `match` has strict semantics for every subfield (undefined -> select all, all subfields must be satisfied in order to match). There is nothing that prevents spec.match.gcp.*, but I would strongly suggest not using it for consistency's sake.

Reply:
Hmm, so would you propose that we un-nest the `gcp` component and actually implement `target` and `exclude` as two separate matchers?

Reply:
That would be ideal. If `target` is empty, are all resources matched? That would be a pre-requisite for un-nesting.

If not, or these two fields should always go together (ex. if `exclude` has no meaning without `target`), it might be worth keeping this as a separate sub-object, but with a more descriptive name than `gcp` (`orgTreeMatcher`, something shorter?), so people know the expected behavior of the matching filter.

Reply:
> If target is empty, are all resources matched?

Yes.

> if exclude has no meaning without target

I *think* exclude could actually be used without target. Like, if I had a label selector I might also want to (separately) exclude a certain folder from that selector. @t12g Thoughts?

Reply:
That seems like a plausible case. So, we could potentially mix and match the 3 matchers then?

label selector + target == only constraining resources with label under target location
label selector + exclude == only constraining resources with label outside of exclude location
label selector + target + exclude == only constraining resources with label under target location, but ignoring those under exclude location

Reply:
Yeah, that makes sense to me. Should we call it `gcpTarget` and `gcpExclude`?

Reply:
I opened #42 to track updating this.

Reply:
The combination of selectors jives with the syntax of the `match` field.

GCP prefixes should be unnecessary. In the event `match` becomes multi-target, we'll need to migrate the CRD schema to use something different.

Reply:
Hmm... Do we need multi-target constraints with KCC? Essentially the same constraint would apply to both CAI resource format and Kubernetes resource format?

Reply:
Yes, I expect that to be come a concern fairly soon. @maxsmythe, can we try to push along the possibility of multi-target constraints?